mindmeld/core.py

```diff
@@ -699,25 +700,75 @@ def from_query(
 
         """
 
-        def _get_form_details(query_span, offset, form_in, form_out):
-            span_out = query.transform_span(query_span, form_in, form_out)
-            full_text = query.get_text_form(form_out)
-            text = span_out.slice(full_text)
+        def _get_token_start(full_norm_text, span_out):
+            """ Calculate the start of a token using the normalized tokens
+            combined as a string and delimited by space.
+
+            Args:
+                full_norm_text (str): Normalized tokens combined by space as a single string.
+                span_out (Span): Span of the token of interest in the full_norm_text.
+            Return:
+                tok_start (int): The starting token index.
+            """
             # The span range is till the span_out or max to the second last char
             tok_start = 0
-            span_range = min(span_out.start, len(full_text) - 1)
-            for idx, current_char in enumerate(full_text[:span_range]):
+            span_range = min(span_out.start, len(full_norm_text) - 1)
+            for idx, current_char in enumerate(full_norm_text[:span_range]):
                 # Increment the counter only if a whitespace follows a non-whitespace
-                next_char = full_text[idx + 1]
+                next_char = full_norm_text[idx + 1]
                 if not current_char.isspace() and next_char.isspace():
                     tok_start += 1
```

Comment:
It's still a little worrisome to me that we are depending on whitespace to mark tokenization, even in the normalized text. If someone's custom tokenizer decides "85 %" is a single token or that "co-op" is a single token normalized as "co op", that shouldn't break the whole system. Perhaps we should switch to using a character that is already system-reserved, like two pipes `||` to distinguish it from the entity annotation.